coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -394,36 +380,46 @@ static boolean isUnionRecordSchemaCompatible(@Nonnull Schema leftSchema, @Nonnul
     for (Schema.Field field : leftSchemaFields) {
       if (!rightSchemaFieldsMap.containsKey(field.name())) {
         // field in leftSchema is missing in rightSchema
-        LOG.error(field.name() + " is in schema " + leftSchema.getName() + ": " + leftSchema.toString(true)
-            + ", but not in schema " + rightSchema.getName() + ": " + rightSchema.toString(true));
-        return false;
+        throw new RuntimeException(
+            field.name() + " is in schema " + leftSchema.getName() + ": " + leftSchema.toString(true)
+                + ", but not in schema " + rightSchema.getName() + ": " + rightSchema.toString(true));
       }
     }
 
     for (Schema.Field field : rightSchemaFields) {
       if (!leftSchemaFieldsMap.containsKey(field.name())) {
         // field in rightSchema is missing in leftSchema
-        LOG.error(field.name() + " is in schema " + rightSchema.getName() + ": " + rightSchema.toString(true)
-            + ", but not in schema " + leftSchema.getName() + ": " + leftSchema.toString(true));
-        return false;
+        throw new RuntimeException(
+            field.name() + " is in schema " + rightSchema.getName() + ": " + rightSchema.toString(true)
+                + ", but not in schema " + leftSchema.getName() + ": " + leftSchema.toString(true));
       }
     }
 
+    List<Schema.Field> mergedSchemaFields = new ArrayList<>();
+
     for (Schema.Field leftField : leftSchemaFields) {
       Schema.Field rightField = rightSchemaFieldsMap.get(leftField.name());
-      boolean isUnionSchemaCompatible = isUnionSchemaCompatible(leftField.schema(), rightField.schema(), strictMode);
-
-      if (!isUnionSchemaCompatible) {
-        LOG.error(leftField.name() + " is not compatible with " + rightField.name() + " for LogicalUnion operator.");
-
-        return false;
+      Schema unionFieldSchema = getUnionFieldSchema(leftField.schema(), rightField.schema(), strictMode);
+      if (unionFieldSchema == null) {
+        unionFieldSchema = getUnionFieldSchema(rightField.schema(), leftField.schema(), strictMode);
+      }
```

Comment:
Because current code and 

> Why do we think `strictMode` should affect the behavior here? Should not the output type be defined this way regardless of the mode?

Thank you for your review, I will remove `strictMode` for this behavior.

Reply:
I am going to add 
```
if (isUnionOrNull(rightSchema) && !isUnionOrNull(leftSchema)) {
      return getUnionFieldSchema(rightSchema, leftSchema, strictMode);
}

  private static boolean isUnionOrNull(Schema schema) {
    return schema.getType() == Schema.Type.UNION || schema.getType() == Schema.Type.NULL;
  }
```
at the beginning of `getUnionFieldSchema`, so that we don't need to call twice and it can still keep `leftSchema` as a base for convenience.